core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -183,7 +190,15 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
         case Code.OK =>
           val updatedLeaderAndIsr = leaderAndIsrs(partition).withZkVersion(setDataResponse.stat.getVersion)
           successfulUpdates.put(partition, updatedLeaderAndIsr)
-        case Code.BADVERSION => updatesToRetry += partition
+        case Code.BADVERSION =>
+          val checkSucceed = setDataResponse.checkResult match {
+            case Some(checkResult) => checkResult.resultCode == Code.OK
+            case None => false
+          }
+          if (checkSucceed)
+            updatesToRetry += partition
+          else
+            failed.put(partition, setDataResponse.resultException.get)
```

Comment:
Thanks for the suggestion. It is simpler and cleaner this way.

Btw, I think we can just use `ControllerMoveException()` in this case. 

Also, related to your comment on my last commit, instead of putting the exception in some data structures, can we just simply throw the exception when the check fails? In this case, we can skip processing unnecessary controller event until we hit `ControllerChange` event. To optimize it further, we can also catch 'ContollerMoveException' explicitly in the `ControllerEventThread` and let the controller resigns immediately.